### PR TITLE
circt: add shared libs

### PR DESCRIPTION
**Pull request by pineapplehunter**
This changes adds circt's shared libs.

Previous circt only included some binaries related to circt, but with this change it will include shared libs and some files from LLVM used by circt. In the future, it might be nice to split the files into smaller packages (e.g. binaries, libs, llvm stuff, ...) but for now I think this should be all right.

Previous version

```bash
$ find -L result | wc -l
10
$ du -shL result
276M	result
```

New version

```bash
$ find -L result | wc -l
5229
$ du -shL result
450M	result
```

CC: @sequencer 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sequencer** [comment]
Ask for review by @SharzyL

File: pkgs/development/compilers/circt/default.nix [hunk 1]
```diff
@@ -62,8 +63,8 @@ stdenv.mkDerivation rec {
 
   installPhase = ''
     runHook preInstall
-    mkdir -p $out/bin
-    mv bin/{{fir,hls}tool,circt-{as,dis,lsp-server,opt,reduce,translate}} $out/bin
+    mkdir -p $out
+    CMAKE_INSTALL_PREFIX=$out cmake --build . --target install --config Release
     runHook postInstall
   '';
```
Comment (NickCao) [inline pkgs/development/compilers/circt/default.nix:69]: Can we just drop installPhase and let cmake do the install automatically?

**Review by NickCao — COMMENTED** [review]
Also try adding a `lib` output to see if upstream build system already supports installing libraries into a different prefix.

**Comment by SharzyL** [comment]
The default install contains too many binaries and libraries from LLVM. I am trying multiple outputs.

File: pkgs/development/compilers/circt/default.nix (hunk 1 above)
Comment (pineapplehunter) [reply pkgs/development/compilers/circt/default.nix:69]: Thanks for the review!
I just tried removing the installPhase, and it seems to work.

**Comment by SharzyL** [comment]
> The default install contains too many binaries and libraries from LLVM. I am trying multiple outputs.

I set `outputs = [ "bin" "lib" "dev" "out" ];` but nix complains
```
error: cycle detected in build of '/nix/store/jhv4qsdi5jm6my2r8lxkmw7lbkdaijh3-circt-1.56.1.drv' in the references of output 'bin' from output 'dev'
```

Trying to figure out why

**Comment by sequencer** [comment]
any progresses? @SharzyL

**Comment by SharzyL** [comment]
After investigation, `dev` contains lot of reference in `$bin$`, and `bin/llvm-config` contains a reference of `$dev`, thus causing a cycling references. A solution is to remove `llvm-config` (and other unnecessary binaries).

I would proposes the following changes
```patch
diff --git a/pkgs/development/compilers/circt/default.nix b/pkgs/development/compilers/circt/default.nix
index f2939ceb53f0..38bd10575332 100644
--- a/pkgs/development/compilers/circt/default.nix
+++ b/pkgs/development/compilers/circt/default.nix
@@ -61,13 +61,6 @@ stdenv.mkDerivation rec {
     substituteInPlace cmake/modules/GenVersionFile.cmake --replace "unknown git version" "${src.rev}"
   '';
 
-  installPhase = ''
-    runHook preInstall
-    mkdir -p $out
-    CMAKE_INSTALL_PREFIX=$out cmake --build . --target install --config Release
-    runHook postInstall
-  '';
-
   doCheck = true;
   checkTarget = "check-circt check-circt-integration";
 
@@ -75,6 +68,17 @@ stdenv.mkDerivation rec {
     rev-prefix = "firtool-";
   };
 
+  # Only keep circt-related binaries in $bin/bin
+  # Also note that bin/llvm-config may contain $dev string in its content, causing cycle references between $bin and $dev
+  postInstall = ''
+    mkdir bin-backup
+    mv $bin/bin/{fir,hls}tool $bin/bin/circt-* bin-backup/
+    rm $bin/bin/*
+    mv bin-backup/* $bin/bin
+  '';
+
+  outputs = [ "bin" "lib" "dev" "out" ];
+
   meta = {
     description = "Circuit IR compilers and tools";
     homepage = "https://circt.org/";
```

Now we have
```
$ du -shL result-*
276M    result-bin
94M     result-dev
81M     result-lib
648M    result-out
```

(where `result-lib` contains only CIRCT libs and `result-out` contains all libs)

**Comment by SharzyL** [comment]
> Or can we move llvm-config to the dev output? I think the same has been done on llvm.

https://github.com/NixOS/nixpkgs/blob/fccf04b6200b711977dbfe9a59e3356bfe768932/pkgs/development/compilers/llvm/15/llvm/default.nix#L379

Oh that's true. Now we can write `postInstall` as
```sh
mkdir -p $dev/bin bin-backup
mv $bin/bin/{fir,hls}tool $bin/bin/circt-* bin-backup
mv $bin/bin/* $dev/bin
mv bin-backup/* $bin/bin
```

But it may make `$dev` much larger, I doubt if circt-dev users need them.

**Comment by sequencer** [comment]
any progress for this PR?

**Comment by pineapplehunter** [comment]
Sorry, I was busy for a while. I'll check and updated this PR tomorrow.

**Comment by pineapplehunter** [comment]
> Or can we move llvm-config to the dev output? I think the same has been done on llvm.

> But it may make $dev much larger, I doubt if circt-dev users need them.

Before I updated this, I'm checking if I can separate circt and llvm.

File: pkgs/development/compilers/circt/default.nix [hunk 2]
```diff
@@ -60,13 +61,17 @@ stdenv.mkDerivation rec {
...
-    mkdir -p $out/bin
-    mv bin/{{fir,hls}tool,circt-{as,dis,lsp-server,opt,reduce,translate}} $out/bin
-    runHook postInstall
+  # Only keep circt-related binaries in $bin/bin
+  # Also note that bin/llvm-config may contain $dev string in its content, causing cycle references between $bin and $dev
+  postInstall = ''
+    mkdir -p $dev/bin bin-backup
+    mv $bin/bin/{fir,hls}tool $bin/bin/circt-* bin-backup
+    mv $bin/bin/* $dev/bin
+    mv bin-backup/* $bin/bin
```
[the remark this answers is not in the document]
Comment (pineapplehunter) [reply pkgs/development/compilers/circt/default.nix:70]: I am not familiar with `moveToOutput`. Are there any docs I can check out?

[the remark this answers is not in the document]
Comment (NickCao) [reply pkgs/development/compilers/circt/default.nix:70]: You can look at `pkgs/build-support/setup-hooks/multiple-outputs.sh` and grep for in-tree uses.

[the remark this answers is not in the document]
Comment (sequencer) [reply pkgs/development/compilers/circt/default.nix:70]: any progress here?

**Comment by pineapplehunter** [comment]
Sorry for not making so much progress.
I have worked on this for a while and it seems that splitting the llvm build produces better results.
I will open a new PR for that.

**Pull request closed without merging**